python: generate Pydantic v2 + typing complete code

multani opened the pull request:
was: `python: Generate fully described Python types`

> **Warning**
>
> Sorry, it's a rather large change, but I think it all helps to produce better type information.
>
> I'm looking for feedback to know if this is the good way to integrate this kind of change, and if this has a chance of being merged.
>
> If needed, I can try to make this big change shorter by splitting this into multiple PRs; it may be harder to see the big picture with smaller PRs though.

---

The goal of this change is to replace the mypy incompatible types (`conlist`, `conint`, `constr`, etc.) by their more modern versions using `Annotated` types with Pydantic v2.


To do so, the code generator changes in several ways:

* The generator introduces several new inner classes to help the code generation:
  - `PythonType` is a recursive object that hold the information about a particular type that the generator is trying to build
  - `PydanticType`, directly translated from the previous `getPydanticType` methods, creates `PythonType` instances from the OpenAPI specifications and produces the final Python types as strings.
  - An initial `Imports` class holds all the new Python's `import` statements that the generator produces along the way.

Via these new classes, the code generation gets improved by the following:

* The string-based formater approach is now replaced by a tree of `PythonType` objects:
  - Each object can hold all its information: type constraints, metadata, etc.
  - By being a tree of itself, types can become naturally nested and allow to easily express complex types
* Most of the code duplication between the 2 `getPydanticType` methods have been unified to remove duplication:
  - Each OpenAPI type definition has a matching `OpenAPI type -> Python type` translation method
  - Common type translations between `CodegenParamerer` and `CodegenProperty` have been unified
* The new `Imports` class can be passed between calls to accumulate all the imports:
  - This removes several flags that were present only to track if something needed to be imported or not.
  - An instance of the class can be shared across multiple method calls to accumulate all the imports that have to be done
  - It reduces the amount of objects to carry around as it can hold any kind of imports.
  - Ultimately, it also allows to control better which Python resources are imported from which modules, and should help to reduce the number of unused imports in the generated code.

References I used to implement the new types:

* The parent issue https://github.com/OpenAPITools/openapi-generator/issues/15464
* A related issue in Pydantic, when Pydantic introduce support for `Annotated`: https://github.com/pydantic/pydantic/issues/975)
* (I will add more as I review the code)

Related:
* Initial fix for https://github.com/OpenAPITools/openapi-generator/issues/15464
* Replacement for https://github.com/OpenAPITools/openapi-generator/pull/15878
* #16384

Note that this improves _only_ a subset of the code generator, but there are still several places that could be improved:

* I think once generalized, usage of the `Imports` class instead of the numerous `Set` instances could help to track all the import dependencies, reduce the generated code (and unused imports) and the code generator itself.
* A whole part of the code generator has not been touched and could benefit from the same kind of improvements.


<!-- Please check the completed items below -->
### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming 7.1.0 minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

@krjakbrjak as Python tech-comittee
@wing328 as the previous major editor of the code generator

multani commented on samples/client/echo_api/python/openapi_client/api/body_api.py:
```diff
@@ -480,7 +480,7 @@ def test_body_multipart_formdata_array_of_binary_with_http_info(self, files : co
 
     @validate_arguments
-    def test_echo_body_free_form_object_response_string(self, body : Annotated[Optional[Dict[str, Any]], Field(description="Free form object")] = None, **kwargs) -> str:  # noqa: E501
+    def test_echo_body_free_form_object_response_string(self, body : Optional[Dict[str, Any]] = None, **kwargs) -> str:  # noqa: E501
```
The removal of Annotated + the description is a regression I need to fix

multani commented:
I removed the `poetry.lock` files I added during my experiment, and it divided the size of the pull request by 2 :D

wing328 commented:
cc @OpenAPITools/generator-core-team

robertschweizer commented on samples/client/echo_api/python/openapi_client/models/pet.py:
```diff
@@ -20,20 +20,21 @@
     """
     id: Optional[StrictInt] = None
-    name: StrictStr = Field(...)
+    name: StrictStr
```
`= Field(...)` was added on purpose, we should not remove it. See https://github.com/OpenAPITools/openapi-generator/pull/15290

robertschweizer replied:
I think passing `...` to `Field()` is required for all fields that can be `None`, but are required (`None` is not the default value). Also in request function signatures.

multani replied:
I will check how to add a test for this, if there's not one yet already :+1:

multani replied:
@robertschweizer  IIRC, this is only a typing issue, not a runtime issue, right?

If yes, I'd propose to let this aside from the time being and mark this as "TODO to improve typing support for the Pydantic v2-compatible generator". Even with all the changes I did here, there's still a long way for a proper typing support, and I guess it will be part of it.

I'll note your comment somewhere though so to not forget it in the new Pydantic v2 generator.

multani replied:
@robertschweizer Correct me if I'm wrong, but I think the new format should be fine with Pydantic v2:

```python
from typing import Optional

from pydantic import BaseModel, StrictStr


class Model(BaseModel):
    required_field: StrictStr
    required_nullable_field: Optional[StrictStr] = None


m1 = Model(required_field="foo", required_nullable_field="bar")

print(m1.required_field.lower())
if m1.required_nullable_field is not None:
    print(m1.required_nullable_field.lower())


m2 = Model(required_field="m2")

print(m2.required_field.lower())
if m2.required_nullable_field is not None:
    print(m2.required_nullable_field.lower())
```
prints:

```
m1
m1 nullable
m2
```

`mypy --strict` says:
```
Success: no issues found in 1 source file
```

pylint says nothing too (only missing docstrings).

If you have another specific example in mind, could you share it and I can take a look!

robertschweizer commented on samples/openapi3/client/petstore/python-aiohttp/petstore_api/models/color.py:
```diff
@@ -30,11 +32,11 @@ class Color(BaseModel):
     """
     # data type: List[int]
-    oneof_schema_1_validator: Optional[conlist(conint(strict=True, le=255, ge=0), max_items=3, min_items=3)] = Field(None, description="RGB three element array with values 0-255.")
+    oneof_schema_1_validator: Optional[Annotated[List[Annotated[int, Field(le=255, strict=True, ge=0)]], Field(min_items=3, max_items=3)]] = Field(default=None, description="RGB three element array with values 0-255.")
```
I do not get an error if I construct this with `oneof_schema_1_validator=[1, 2]`. Maybe pydantic is thrown off by `Field()` being used on both sides of `=`? Also pydantic might not take into account `Field()` restrictions in the first argument of `Annotated[]`.

Best add a unit-test to make sure validation works, or is there one already?

multani replied:
I will check this :+1:

multani replied:
I added more tests: this works with Pydantic v2 but not with Pydantic v1. We'll focus on the former anyway and I'll merge my PR in a Pydantic v2-only generator :+1:

robertschweizer replied:
This is great, thanks a lot for moving forward to pydantic v2! Then things work of course :)

robertschweizer reviewed (commented):
Thanks a lot for the PR! I checked some of the generated classes.

multani commented:
Thanks to the merge of https://github.com/OpenAPITools/openapi-generator/pull/16643/, I have plenty of test failures now :)

I will turn this back to draft while I'm fixing the comments and the tests :+1:

multani commented:
I did some testing locally to verify some constructions, but I just realized I was using Pydantic v2 during my tests, whereas all the codebase is still using Pydantic v1.

Some constructions are not working correctly, for instance, I rewrote:

```python
class XXX(BaseModel):
    field: Optional[
        conlist(
            conint(strict=True, le=255, ge=0),
            max_items=3, min_items=3,
        )
    ]
```
As:
```python
class XXX(BaseModel):
    field: Optional[
        Annotated[
            List[Annotated[int, Field(strict=True, le=255, ge=0)]],
            Field(max_items=3, min_items=3),
        ]
    ]

```

Using the following code:
```python
test = XXX(field=[0, 0, 0, 0])
print(test)
```

In Pydantic v1, the original construction with `conlist` fails with:

```
pydantic.error_wrappers.ValidationError: 1 validation error for XXX
field
  ensure this value has at most 3 items (type=value_error.list.max_items; limit_value=3)
```

But the new construction passes silently:

```
field=[0, 0, 0, 0]
```

In Pydantic v2, the new construction fails with:
```
pydantic_core._pydantic_core.ValidationError: 1 validation error for XXX
field
  List should have at most 3 items after validation, not 4 [type=too_long, input_value=[0, 0, 0, 0], input_type=list]
    For further information visit https://errors.pydantic.dev/2.3/v/too_long
```

There's already some work undergoing to use Pydantic v2 in https://github.com/OpenAPITools/openapi-generator/pull/16636.

It's hard to know right now how much this will conflict with what I'm trying to do in this pull request, so:

1. I'm going to wait for the Pydantic v2 pull request to be merged first
2. I'll check to rebase my work on top on the Pydantic v2 change, then address the failures

In the meantime, I guess I'll put this on hold.

wing328 commented:
@multani is there anything we can help to get this enhancement for pydantic 1.x (used by current python client generator as well) as well?

I suggest we get this merged first and then we can start working on supporting both pydantic 1.x and 2.x

wing328 commented:
In the coming week, please kindly PM me via Slack when you've time: 

https://join.slack.com/t/openapi-generator/shared_invite/zt-12jxxd7p2-XUeQM~4pzsU9x~eGLQqX2g

(I agree putting this PR on hold for the time being)

Have a nice weekend.

multani commented:
This now generates a mostly Pydantic v2 compatible code. There are a few things that I didn't address here:

* I completely ignored the warnings for deprecated APIs so there are several Pydantic APIs that still need to be changed: the validator decorators, `min_length` vs instead of `min_items`, etc.

There are a couple of things that still don't work:

### Circular dependencies

The "circular dependency" test (`CircularReferenceModel` imports `FirstRef`, which imports `SecondRef`, which imports `CircularReferenceModel`) in [`petstore_api/models/circular_reference_model.py`](https://github.com/multani/openapi-generator/blob/python-typing-1/samples/openapi3/client/petstore/python/petstore_api/models/circular_reference_model.py) totally doesn't work :facepalm:
This seems to be more a Python/Pydantic issue though, see:

* https://github.com/pydantic/pydantic/issues/707
* https://github.com/pydantic/pydantic/issues/1873

To be honest, I don't know how it even worked before :)

This is solvable, if we bundle all the circular dependencies in the same module, and then use Pydantic `model_rebuild` function, see https://docs.pydantic.dev/2.3/usage/postponed_annotations/#cyclic-references

### float

It took me a while to figure this one, but [`format_test.py`](https://github.com/multani/openapi-generator/blob/python-typing-1/samples/openapi3/client/petstore/python/petstore_api/models/format_test.py#L34-L36) contains something like this:

```python
from pydantic import BaseModel

class FormatTest(BaseModel):
    number: float
    float: Optional[float] = None
```

Then, running this:

```python
print(FormatTest(number=3.2))
```

fails with:

```
pydantic_core._pydantic_core.ValidationError: 1 validation error for FormatTest
number
  Input should be None [type=none_required, input_value=3.2, input_type=float]
    For further information visit https://errors.pydantic.dev/2.3/v/none_required
```

It took me a while to understand, but the `float` type gets aliases to `None` :)

I'm not sure how to fix this at the moment, I commented the related code. Note that simply **loading** a model which contains such a definition fails all the tests.


### `actual_instance`

There are still some test failures related to `actual_instance`, for instance in [`test_model.py`](https://github.com/multani/openapi-generator/blob/b3f7d7da9dc466851b07f3f1e0a07c28037b547a/samples/openapi3/client/petstore/python/tests/test_model.py#L135-L143).

This is probably fixable but I don't understand exactly the purpose of the code here and the tests are a bit weird (`from_dict("a")`?)


### Other failures

There are a few other failures, but I think they should be solvable.


### About typing itself

I also had a look at the output of mypy in the samples `echo_client`, and I think mypy should be not too far to validate almost all the code once we finish the migration (removing the Pydantic v2 warnings, etc.)

I'm not too familiar with all the features of the generated code, but it feels there are also opportunities to simplify the generated code by *not* generating part of it, or by reusing what Pydantic offers builtin now (`__properties` list attributes, custom implementation of `to_json()`, `to_dict()` and related `from_*()` methods, etc.)

I think this can be done in a second step.

multani commented:
> * I completely ignored the warnings for deprecated APIs so there are several Pydantic APIs that still need to be changed: the validator decorators, `min_length` vs instead of `min_items`, etc.

This is still not done, but the code seems to be working and this could be done in a 2nd step.

> There are a couple of things that still don't work:
> ### Circular dependencies
> 
> The "circular dependency" test (`CircularReferenceModel` imports `FirstRef`, which imports `SecondRef`, which imports `CircularReferenceModel`) in [`petstore_api/models/circular_reference_model.py`](https://github.com/multani/openapi-generator/blob/python-typing-1/samples/openapi3/client/petstore/python/petstore_api/models/circular_reference_model.py?rgh-link-date=2023-09-24T19%3A16%3A18Z) totally doesn't work 🤦 This seems to be more a Python/Pydantic issue though, see:
> 
>     * [Models referencing each other from separate files pydantic/pydantic#707](https://github.com/pydantic/pydantic/issues/707)
> 
>     * [How to avoid circular imports when using models (defined in different modules) which refers to each other? pydantic/pydantic#1873](https://github.com/pydantic/pydantic/issues/1873)
> 
> 
> To be honest, I don't know how it even worked before :)
> 
> This is solvable, if we bundle all the circular dependencies in the same module, and then use Pydantic `model_rebuild` function, see [docs.pydantic.dev/2.3/usage/postponed_annotations/#cyclic-references](https://docs.pydantic.dev/2.3/usage/postponed_annotations/#cyclic-references)

I "fixed" this by *not calling* `Model.model_rebuild()` in the code. For now, the typing won't fully work for these dependent modules.
I added a `TODO: pydantic v2` in the code to re-enable this.

> ### float
> 
> It took me a while to figure this one, but [`format_test.py`](https://github.com/multani/openapi-generator/blob/python-typing-1/samples/openapi3/client/petstore/python/petstore_api/models/format_test.py?rgh-link-date=2023-09-24T19%3A16%3A18Z#L34-L36) contains something like this:
> 
> ```python
> from pydantic import BaseModel
> 
> class FormatTest(BaseModel):
>     number: float
>     float: Optional[float] = None
> ```
> 
> Then, running this:
> 
> ```python
> print(FormatTest(number=3.2))
> ```
> 
> fails with:
> 
> ```
> pydantic_core._pydantic_core.ValidationError: 1 validation error for FormatTest
> number
>   Input should be None [type=none_required, input_value=3.2, input_type=float]
>     For further information visit https://errors.pydantic.dev/2.3/v/none_required
> ```
> 
> It took me a while to understand, but the `float` type gets aliases to `None` :)
> 
> I'm not sure how to fix this at the moment, I commented the related code. Note that simply **loading** a model which contains such a definition fails all the tests.

I "fixed" this by commenting the `float` inside the OpenAPI YAML file :)

This is fixable, but in the general case, it's a bit hard:

* Everytime the generator generates a type (or subtype) for a model property:
  * We need to check if there's another model property with the same name as the type we are trying to generate:
    * If no, there it's all good
    * If a model property exist with the same name as the type (or subtype), we need to:
      * Create an alias for the type, for instance `Float = float`
      * The alias must also not be another model property
      * Then substitute the actual type with the alias
      * Generate the list of all the aliases before generating the model

It's not impossible, but a bit of work. I deferred that to later, there are `TODO: pydantic v2` in the code.

> ### `actual_instance`
> 
> There are still some test failures related to `actual_instance`, for instance in [`test_model.py`](https://github.com/multani/openapi-generator/blob/b3f7d7da9dc466851b07f3f1e0a07c28037b547a/samples/openapi3/client/petstore/python/tests/test_model.py#L135-L143).
> 
> This is probably fixable but I don't understand exactly the purpose of the code here and the tests are a bit weird (`from_dict("a")`?)
> ### Other failures
> 
> There are a few other failures, but I think they should be solvable.

This is fixed now. I was using [Pydantic's `.model_dump_json()`](https://docs.pydantic.dev/latest/concepts/serialization/#modelmodel_dump_json), but it didn't play well with how objects are currently composed.

I think the generated could be simplified and improved by reusing more Pydantic methods instead of implementing our own.

This could also be improved later on.

> ## About typing itself
> 
> I also had a look at the output of mypy in the samples `echo_client`, and I think mypy should be not too far to validate almost all the code once we finish the migration (removing the Pydantic v2 warnings, etc.)
> 
> I'm not too familiar with all the features of the generated code, but it feels there are also opportunities to simplify the generated code by _not_ generating part of it, or by reusing what Pydantic offers builtin now (`__properties` list attributes, custom implementation of `to_json()`, `to_dict()` and related `from_*()` methods, etc.)
> 
> I think this can be done in a second step.

I think this could be looked at again once we have fixed all (most of) the warnings for Pydantic v2. There are still a lot of typing errors, but more than 60% should go away once the migration has progressed further.

multani commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractPythonCodegen.java:
```diff
@@ -1065,624 +1060,103 @@ private String getPydanticType(CodegenParameter cp,
-            if (cp.hasValidation) {
-                List<String> fieldCustomization = new ArrayList<>();
-                List<String> intFieldCustomization = new ArrayList<>();
+    public String toEnumVariableName(String name, String datatype) {
```
This is from old version, line 1615

multani commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractPythonCodegen.java:
```diff
@@ -1065,624 +1060,103 @@ private String getPydanticType(CodegenParameter cp,
+        return pt.getType(cp);
+    }
+
+    public void setMapNumberTo(String mapNumberTo) {
```
This is form old version, line 1605

multani commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractPythonCodegen.java:
```diff
@@ -1065,624 +1060,103 @@ private String getPydanticType(CodegenParameter cp,
-            LOGGER.warn("Codegen property is null (e.g. map/dict of undefined type). Default to typing.Any.");
-            typingImports.add("Any");
-            return "Any";
+    void createImportMapOfSet(String modelName, Map<String, CodegenModel> codegenModelMap) {
```
This is from old version, line 1660

multani commented on modules/openapi-generator/src/main/resources/python/api_client.mustache:
```diff
@@ -302,6 +302,9 @@ class ApiClient:
         elif isinstance(obj, list):
             return [self.sanitize_for_serialization(sub_obj)
                     for sub_obj in obj]
+        elif isinstance(obj, set):
+            return {self.sanitize_for_serialization(sub_obj)
+                    for sub_obj in obj}
```
Not useful for now. `Annotated[List, Field(unique_items=True)]` should be replaced by `Annotated[Set]`, but this has further implications (sets lose the ordering, not serializable as JSON automatically, etc.)

```suggestion
```

wing328 replied:
thanks for the heads-up. later we can add some tests in https://github.com/OpenAPITools/openapi-generator/blob/master/samples/openapi3/client/petstore/python/tests/test_api_client.py#L159 to cover `set`

multani replied:
Actually, I wanted to remove this block of code, as it's unused right now.
There's a comment in the related part of the Java code:

```java
        private PythonType arrayType(IJsonSchemaValidationProperties cp) {
            PythonType pt = new PythonType();
            if (cp.getMaxItems() != null) {
                pt.constrain("max_items", cp.getMaxItems());
            }
            if (cp.getMinItems()!= null) {
                pt.constrain("min_items", cp.getMinItems());
            }
            if (cp.getUniqueItems()) {
                // A unique "array" is a set
                // TODO: pydantic v2: Pydantic suggest to convert this to a set, but this has some implications:
                // https://github.com/pydantic/pydantic-core/issues/296
                // Also, having a set instead of list creates complications:
                // random JSON serialization order, unable to easily serialize
                // to JSON, etc.                                                                                                                                                                                                                                                
                //pt.setType("Set");
                //typingImports.add("Set");
                pt.setType("List");
                typingImports.add("List");
            } else {
                pt.setType("List");
                typingImports.add("List");
            }
            pt.addTypeParam(getType(cp.getItems()));
            return pt;
        }
```

multani commented on modules/openapi-generator/src/main/resources/python/model_anyof.mustache:
```diff
@@ -23,17 +24,14 @@ class {{classname}}({{#parent}}{{{.}}}{{/parent}}{{^parent}}BaseModel{{/parent}}
-    else:
-        actual_instance: Any
-    any_of_schemas: List[str] = Field({{#lambda.uppercase}}{{{classname}}}{{/lambda.uppercase}}_ANY_OF_SCHEMAS, const=True)
+    actual_instance: Optional[Union[{{#anyOf}}{{{.}}}{{^-last}}, {{/-last}}{{/anyOf}}]] = None
+    one_of_schemas: List[str] = Literal[{{#lambda.uppercase}}{{{classname}}}{{/lambda.uppercase}}_ANY_OF_SCHEMAS]
```
```suggestion
    if TYPE_CHECKING:
        actual_instance: Optional[Union[{{#anyOf}}{{{.}}}{{^-last}}, {{/-last}}{{/anyOf}}]] = None
    else:
        actual_instance: Any = None
    any_of_schemas: List[str] = Literal[{{#lambda.uppercase}}{{{classname}}}{{/lambda.uppercase}}_ANY_OF_SCHEMAS]
```

multani commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractPythonCodegen.java:
```diff
@@ -1994,4 +1434,799 @@ public String toEnumDefaultValue(String value, String datatype) {
+            }
+
+            // field
+            if (cp.baseName != null && !cp.baseName.equals(cp.name)) { // base name not the same as name
+                pt.annotate("serialization_alias", cp.baseName);
+            }
```
This used to be `alias`, but in Pydantic v2, [`alias` is both a validator and a serializer](https://docs.pydantic.dev/2.3/usage/fields/#field-aliases).

From the tests, I assume `alias` was used for serialization only (to produce the aliased field), not for validation (we still want to be able to read the original Pythonic field)

Note that in Pydantic v1, both forms (the field name and the alias name) could be used to create a new instance of the model. This is not possible anymore, only the field name is accepted.

wing328 commented on modules/openapi-generator/src/main/resources/python/api.mustache:
```diff
@@ -7,9 +7,6 @@ import io
 from pydantic import validate_arguments, ValidationError
-{{#asyncio}}
-from typing import overload, Optional, Union, Awaitable
-{{/asyncio}}
```
why these lines are removed?

multani replied:
2 reasons:

* These imports are not asyncio-specific
* The imported names are not used in the template

The names could be used by the code generated by the Java code, in which case the imports would be provided from the Java code instead.

wing328 replied:
👌

wing328 commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractPythonCodegen.java:
```diff
@@ -1009,19 +997,27 @@ private ModelsMap postProcessModelsMap(ModelsMap objs) {
 
             // import models one by one
-            if (!modelImports.isEmpty()) {
+            {
```
why removed the check here?

multani replied:
I don't remember, I will add it back.

I need to change the place where `x-py-model-imports` (there could be resources to imports even if there no dependencies between models), I will move it outside the `if` block.

multani replied:
I re-added the condition and moved the `otherImports.exports()` imports out of the block.

b0g3r commented on modules/openapi-generator/src/main/resources/python/model_anyof.mustache:
```diff
@@ -24,16 +28,16 @@ class {{classname}}({{#parent}}{{{.}}}{{/parent}}{{^parent}}BaseModel{{/parent}}
 {{/composedSchemas.anyOf}}
     if TYPE_CHECKING:
-        actual_instance: Union[{{#anyOf}}{{{.}}}{{^-last}}, {{/-last}}{{/anyOf}}]
+        actual_instance: Optional[Union[{{#anyOf}}{{{.}}}{{^-last}}, {{/-last}}{{/anyOf}}]] = None
```
@multani Hi Jonathan! Do you remember, by any chance, why do you introduce Optional here? Is it possible to have None value here even with one_of validator?

multani replied:
Hey :wave: Sorry, I don't remember the details of this particular change :/ Does it create a problem for you?

b0g3r replied:
Yes, unfortunately. Before this changes non-optional anyOf had correct typing (Union[A, B]), but now it has additional None in the union, which isn't right and stops me from using proper typing. Should I try to change it back and create PR for it?

multani replied:
Yes, of course, feel free to create a pull request!

I just checked again, and there's also a `is None` check in the `to_dict` method at the end of the file, it's probably why I added the `Optional` type in the first place. If you remove `Optional`, can you also remove that check?
